### PR TITLE
feat: implement standard Bubbles, port x/ansi, and add additional package ports from Charm Go repositories for TUI design

### PR DESCRIPTION
**Pull request by WilliamAGH**
## Summary
This major update aligns TUI4J's architecture 1:1 with the **Charmbracelet ecosystem** (Go). It introduces the `compat.*` top-level package structure mirroring `bubbletea`, `bubbles`, `lipgloss`, and `harmonica`, implements the complete suite of standard Bubbles (components), and integrates a canonical port of `x/ansi` for robust terminal rendering. All architectural changes include backward-compatible deprecation shims to ensure **no breaking changes** for existing consumers.

### Features: Standard Bubbles
- **FilePicker**: Full filesystem navigation with filtering, permission handling (Windows/POSIX), and error surfacing.
- **Progress**: Spring-animated progress bars with solid/gradient colors and percentage views.
- **Table**: Rich table component with customizable columns, rows, styles, and border rendering.
- **Textarea**: Multi-line text editing with cursor navigation, scrolling, and paste support.
- **Viewport**: High-performance content scrolling with mouse wheel support.
- **Stopwatch**: Precision timing with Go-style duration formatting (e.g., "1m30s").

### Core & Infrastructure
- **x/ansi Port**: Integrated `charmbracelet/x/ansi` for correct ANSI parsing, state-machine based stripping, and grapheme-aware width calculation.
- **Exec**: Implemented `ExecProcess` with event loop blocking to match Bubble Tea behavior (fixes concurrency race conditions).
- **Input**: Added **Bracketed Paste** support (emitting `PasteMessage`) and proper SSH environment variable forwarding (`TERM`/`COLORTERM`).

### Architecture & Refactoring
- **Package Alignment**: Restructured code into `compat.bubbletea`, `compat.bubbles`, `compat.lipgloss`, and `compat.harmonica` to mirror the distinct Charmbracelet repositories.
- **Compatibility**: Added `@Deprecated` shims for all relocated classes (e.g., `compat.bubbletea.bubbles.*` -> `compat.bubbles.*`) to maintain API stability.
- **Message Types**: Standardized on canonical `*Message` types with `*Msg` variants available for parity with Charm.

### Examples
- **New Structure**: Reorganized build to use dedicated source sets (`examplesBubbletea`, `examplesBubbles`) matching the new architecture.
- **New Demos**: Added comprehensive examples including `chat` (Textarea+Viewport), `credit-card-form` (Validation), `file-picker`, `table` (Pokemon), `pager`, and `real-time` (Debounce).

### Critical Fixes
- **Concurrency**: Fixed a deadlock in `ExecProcess` by draining process streams concurrently.
- **Stability**: Fixed bounds checking in `List` (empty filters), `Viewport` (page down), and `FilePicker` (empty directories).
- **Windows**: Added permission fallbacks for `FilePicker` on non-POSIX systems.

## Net New Ports (Charm Ecosystem Alignment)
The following repositories and classes are **brand new additions** to the codebase, ported directly from their Go counterparts:

### `charmbracelet/bubbles` (New Components)
| Component | Java Package (`...compat.bubbles`) | Description |
|-----------|------------------------------------|-------------|
| **FilePicker** | `.filepicker.*` | Filesystem navigation, directory traversing, and selection. |
| **Progress** | `.progress.*` | Animated progress bars (spring-based) with gradients. |
| **Stopwatch** | `.stopwatch.*` | Elapsed time tracking with start/stop/reset commands. |
| **Table** | `.table.*` | Data tables with headers, borders, and row selection. |
| **Textarea** | `.textarea.*` | Multi-line text editing, cursor movement, and soft wrapping. |
| **Viewport** | `.viewport.*` | Scrollable window for viewing large content. |

### `charmbracelet/x` (New Infrastructure)
| Package | Java Package (`...compat.x`) | Key Classes |
|---------|------------------------------|-------------|
| **ansi** | `.ansi` | `StringWidth`, `Truncate`, `Cut`, `Strip` |
| **ansi/parser** | `.ansi.parser` | `State`, `Action`, `TransitionTable`, `Parser` |

### `charmbracelet/harmonica` (New Physics)
| Package | Java Package (`...compat.harmonica`) | Key Classes |
|---------|--------------------------------------|-------------|
| **harmonica** | `.harmonica` | `Spring`, `Projectile`, `Point`, `Vector` |

## Technical Details
- **Language**: Java 21
- **Upstream Reference**: Parity with Charmbracelet's `bubbletea` v0.25+, `bubbles` v0.18+, and `lipgloss` v0.10+.
- **Testing**: Added golden file tests for `Table` and `Harmonica` physics.

## Breaking Changes
- **None**: Legacy package paths remain available as deprecated shims extending the new canonical implementations.
- **Deprecations**: `Program.ansiCompressor` is deprecated (matching upstream removal).

### CODE REVIEW

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/bubbles/textarea/Textarea.java [hunk 1]
```diff
@@ -0,0 +1,1462 @@
...
+                break;
+            }
+            setCursor(col - 1);
+        }
+
+        while (col > 0) {
+            if (!UCharacter.isWhitespace(value.get(row)[col])) {
+                setCursor(col - 1);
+            } else {
+                if (col > 0) {
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/bubbles/textarea/Textarea.java]: # Suggestion Implemented ✅

## Changes Made
- [x] Removed redundant `if (col <= 0)` check inside `while (col > 0 ...)` loop.
- [x] Removed redundant `if (col > 0)` check in the second `while (col > 0)` loop (implied by loop condition and logic flow).

## Testing
- [x] Build: Successful (`./gradlew build`)
- [x] Tests: All tests passed

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/lipgloss/Style.java [hunk 2]
```diff
@@ -596,8 +658,50 @@ public Style copy() {
...
+        this.borderTopForeground = style.borderTopForeground;
+        this.borderRightForeground = style.borderRightForeground;
+        this.borderBottomForeground = style.borderBottomForeground;
+        this.borderLeftForeground = style.borderLeftForeground;
+        this.borderTopBackground = style.borderTopBackground;
+        this.borderRightBackground = style.borderRightBackground;
+        this.borderBottomBackground = style.borderBottomBackground;
+        this.borderLeftBackground = style.borderLeftBackground;
         return this;
     }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/lipgloss/Style.java]: # Suggestion Implemented ✅

## Changes Made
- [x] Updated `inherit()` to strictly follow upstream `lipgloss` behavior.
- [x] Now only copies properties if the receiver's property is unset (null/0/false).
- [x] Explicitly ignores padding and margin properties during inheritance.
- [x] Added `testInheritSkipsLayout` to verify correct behavior.

## Testing
- [x] Build: Successful (`./gradlew build`)
- [x] Tests: All tests passed (including new verification test)

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/lipgloss/Style.java [hunk 3]
```diff
@@ -447,6 +490,25 @@ public String render(String... strings) {
...
+            String[] maxWidthLines = string.split("\n");
+            for (int i = 0; i < maxWidthLines.length; i++) {
+                maxWidthLines[i] = Truncate.truncate(maxWidthLines[i], maxWidth, ellipsis);
+            }
+            string = String.join("\n", maxWidthLines);
+        }
+
+        if (maxHeight > 0) {
+            String[] maxHeightLines = string.split("\n");
+            int height = Math.min(maxHeight, maxHeightLines.length);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/lipgloss/Style.java]: # Suggestion Implemented ✅

## Changes Made
- [x] Renamed local variable `height` to `displayHeight` to avoid shadowing the class field.

## Testing
- [x] Build: Successful (`./gradlew build`)
- [x] Tests: All tests passed

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/bubbles/textarea/Textarea.java [hunk 4]
```diff
@@ -0,0 +1,1462 @@
...
+            }
+        }
+
+        while (col > 0) {
+            if (UCharacter.isWhitespace(value.get(row)[col - 1])) {
+                break;
+            }
+            setCursor(col - 1);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/bubbles/textarea/Textarea.java]: # Suggestion Implemented ✅

## Changes Made
- [x] Added boundary check `if (row == 0 && col == 0)` inside the `wordLeft` loop.
- [x] Prevents infinite looping when the document starts with whitespace and the cursor is already at the beginning.

## Testing
- [x] Build: Successful (`./gradlew build`)
- [x] Tests: All tests passed

File: src/main/java/com/williamcallahan/tui4j/compat/lipgloss/Renderer.java [hunk 5]
```diff
@@ -32,6 +31,26 @@ public Renderer(Output output) {
...
+        try {
+            this.output = Output.withEnvironment(environment);
+            // Reset cached values so they're re-detected with new environment
+            this.colorProfile = null;
+            this.explicitColorProfile = false;
+            this.hasDarkBackgroundSet = false;
+        } finally {
+            renderLock.unlock();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/lipgloss/Renderer.java]: # Suggestion Implemented ✅

## Changes Made
- [x] Only reset `colorProfile` if `explicitColorProfile` is false
- [x] Ensures user-set color profile is preserved when environment changes

## Commit Details
- **Files Modified**: `Renderer.java`
- **Type Safety**: Preserved

## Testing
- [x] Compilation passed

File: src/main/java/com/williamcallahan/tui4j/compat/lipgloss/Style.java [hunk 6]
```diff
@@ -447,6 +496,25 @@ public String render(String... strings) {
...
+
+        if (maxHeight > 0) {
+            String[] maxHeightLines = string.split("\n");
+            int displayHeight = Math.min(maxHeight, maxHeightLines.length);
+            if (maxHeightLines.length > 0) {
+                String[] truncatedLines = new String[displayHeight];
+                System.arraycopy(maxHeightLines, 0, truncatedLines, 0, displayHeight);
+                string = String.join("\n", truncatedLines);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/lipgloss/Style.java]: # Suggestion Analysis 🔍

## Decision: Not Implemented

## Analysis Results
- [x] **Current Implementation**: Render logic as written
- [x] **Suggested Change**: MaxWidth/MaxHeight handling
- [x] **Risk Assessment**: `Style` port is complex and this logic matches upstream Go `lipgloss` behavior where these constraints are applied during render.

## Reasoning
The current implementation aligns with how the Go `lipgloss` library handles style rendering and constraints. Diverging from this behavior without a comprehensive test suite covering all edge cases risks introducing subtle layout bugs.

## Conclusion
The current implementation is preferred to maintain parity with the upstream Go library.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbles/timer/TickMessage.java [hunk 7]
```diff
@@ -1,10 +1,14 @@
-package com.williamcallahan.tui4j.compat.bubbletea.bubbles.timer;
+package com.williamcallahan.tui4j.compat.bubbles.timer;
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbles/timer/TickMessage.java]: User confirmed shims were created to address the package structure inconsistencies.

[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbles/timer/TickMessage.java]: # Suggestion Analysis 🔍

## Decision: Not Implemented

## Analysis Results
- [x] **Current Implementation**: Placed in `com.williamcallahan.tui4j.compat.bubbles.timer`
- [x] **Suggested Change**: Move to `com.williamcallahan.tui4j.compat.bubbletea.bubbles.timer`
- [x] **Risk Assessment**: Violates specific Package Mapping in AGENTS.md

## Reasoning
The project's `AGENTS.md` explicitly defines the Package Mapping:
- `bubbletea (core)` -> `com.williamcallahan.tui4j.compat.bubbletea.*`
- `bubbles/*` -> `com.williamcallahan.tui4j.compat.bubbles.*`

The separate `bubbles` package aligns with the upstream separation.

## Conclusion
The current implementation is preferred because it adheres to the defined Package Mapping table.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/Program.java [hunk 8]
```diff
@@ -341,6 +494,81 @@ private Model eventLoop() {
...
+
+    private void handleSequence(Command... commands) {
+        Arrays.stream(commands)
+                .reduce(
+                        CompletableFuture.completedFuture(null),
+                        (CompletableFuture<Void> future, Command command) -> future.thenCompose(
+                                __ -> commandExecutor.executeIfPresent(command, this::send, this::sendError)),
+                        (f1, f2) -> f2)
+                .join();
+    }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/Program.java]: Implemented asynchronous sequence handling using `thenCompose` and monotonic sequence IDs (`long`) to prevent blocking the event loop while ensuring correct message ordering, matching the Bubble Tea behavior.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbles/list/List.java [hunk 9]
```diff
@@ -444,6 +708,12 @@ protected Command resetFiltering() {
         return fetchCurrentPageItems();
     }
 
+    /**
+     * Updates the key bindings based on the current state.
+     */
+    /**
+     * Updates key bindings based on the current state.
+     */
     protected void updateKeybindings() {
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbles/list/List.java]: The duplicate Javadoc seems to have been removed already. The current code shows only one Javadoc block.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbles/paginator/KeyMap.java [hunk 10]
```diff
@@ -18,18 +23,37 @@ public static KeyMap defaultKeyMap() {
     private Binding prevPage;
     private Binding nextPage;
 
+    /**
+     * Creates a new empty key map.
+     */
     public KeyMap() {
     }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbles/paginator/KeyMap.java]: Documented the nullability in the no-arg constructor as suggested.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/Logging.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        public synchronized void setOutput(OutputStream output) {
+            if (handler != null) {
+                logger.removeHandler(handler);
+            }
+            StreamHandler streamHandler = new StreamHandler(output, new PrefixedFormatter(prefix));
+            streamHandler.setLevel(Level.ALL);
+            logger.addHandler(streamHandler);
+            logger.setUseParentHandlers(false);
+            handler = streamHandler;
+        }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/Logging.java]: The suggested improvements (logger level setting and auto-flushing handler) have already been implemented.

File: src/main/java/com/williamcallahan/tui4j/compat/bubbletea/render/Renderer.java [hunk 12]
```diff
@@ -3,102 +3,204 @@
...
+     * <p>
+     * tui4j extension; no Bubble Tea equivalent.
+     */
     void notifyModelChanged();
 
-    // Bubble Tea: seeks to replicate bubbletea/standard_renderer.go handleMessages behavior.
+    /**
+     * Processes a message that might affect the renderer state (e.g. window resize).
+     */
     void handleMessage(Message msg);
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/com/williamcallahan/tui4j/compat/bubbletea/render/Renderer.java]: The @param tag has been added.
